lib/paloaltofw.py

```diff
@@ -0,0 +1,583 @@
+        self.options = {}
+        self.options["from_zone"] = [from_zone]
+        self.options["to_zone"] = [to_zone]
+
+        self.modify_options(terms)
+
+    def modify_options(self, PAFWterm):
+        # for PAFWterm in terms:
+        term = PAFWterm.term
+        self.options["source"] = []
+        self.options["destination"] = []
+        self.options["application"] = []
+        self.options["service"] = []
+        self.options["action"] = "allow"
+
+        # SOURCE-ADDRESS
+        if term.source_address:
+            saddr_check = set()
+            for saddr in term.source_address:
+                saddr_check.add(saddr.parent_token)
+            saddr_check = sorted(saddr_check)
+            for addr in saddr_check:
+                self.options["source"].append(str(addr))
+        else:
+            self.options["source"].append('any')
+
+        # DESTINATION-ADDRESS
+        if term.destination_address:
+            daddr_check = set()
+            for daddr in term.destination_address:
+                daddr_check.add(daddr.parent_token)
+            daddr_check = sorted(daddr_check)
+            for addr in daddr_check:
+                self.options["destination"].append(str(addr))
+        else:
+            self.options["destination"].append('any')
+
+        if term.action:
+            self.options["action"] = term.action[0]
+
+        if term.pan_application:
+            for pan_app in term.pan_application:
+                self.options["application"].append(pan_app)
+
+        if term.destination_port:
+            ports = []
+            for tup in term.destination_port:
+                if len(tup) > 1 and tup[0] != tup[1]:
+                    ports.append(str(tup[0]) + "-" + str(tup[1]))
+                else:
+                    ports.append(str(tup[0]))
+            ports = tuple(ports)
+
+            # check to see if this service already exists
+            for p in term.protocol:
+                if (ports, p) in Service.service_map:
+                    self.options["service"].append(
+                        Service.service_map[(ports, p)]["name"])
+                else:
+                    # create service
+                    new_service = Service(ports, term.name, p)
+                    self.options["service"].append(
+                        Service.service_map[(ports, p)]["name"])
+        if term.protocol:
+            if term.protocol[0] == "icmp":
+                self.options["application"].append("ping")
+
+        rule_name = "-".join(self.options["from_zone"]) + "_2_" + \
+            "-".join(self.options["to_zone"]) + "-" + term.name
+# TODO: uncomment when ready to use
+#    if len(rule_name.decode("utf-8")) > 31:
+#        raise PaloAltoFWTooLongName('Rule name must be 31 characters max: %s'
+#                                        % str(rule_name))
+        self.rules[rule_name] = self.options
+
+
+class PaloAltoFW(aclgenerator.ACLGenerator):
+    """PaloAltoFW rendering class.
```

Comment:
This docstring should only be one line. Arguments should not be documented here and the description below is kind of redundant. It inherits from the aclgenerator class so it should be obvious it renders a syntax output.